keps/2724-topology-aware-schedling/README.md

```diff
@@ -0,0 +1,759 @@
+
+A more involving approach would be to recompute the TopologyAdmission, however,
+until now we don't modify the workload's admission while the workload is
+scheduled, so it would require extra investigation and effort. We will consider
+this before graduation to GA based on investigation if feasible and feedback
+from users.
+
+#### Race condition when accounting for DaemonSet pods
+
+There is a risk that workloads are scheduled before the DaemonSet pods are
+accounted by TAS.
+
+We consider this risk as not that relevant for alpha as the TAS workloads are
+expected to mostly consume accelerators (like TPUs and GPUs), so they are not
+expected to compete for resources with DaemonSet pods.
+
+One way to mitigate the risk is to implement tracking for DaemonSets
+(see [Account usage by watching DaemonSet pods](#account-usage-by-watching-daemonset-pods)).
+We will re-evaluate the need for Beta or GA based on the users' feedback.
+
+## Design Details
+
+### Hierarchy representation
+
+We propose the model for representing the hierarchy of nodes within a data center
+by using node labels. We assume the node labels are set up by a cloud provider,
+or set up manually by administrators of on-premise clusters.
+
+Additionally, we assume that every node used for TAS has a set of the labels
+which identifies uniquely its location in the tree structure. We do not assume
+global uniqueness of labels on each level, i.e. there could be two nodes with
+the same "rack" label, but in different "blocks".
+
+For example, this is a representation of the dataset hierarchy;
+
+|  node  |  cloud.provider.com/topology-block | cloud.provider.com/topology-rack |
+|:------:|:----------------------------------:|:--------------------------------:|
+| node-1 |               block-1              |              rack-1              |
+| node-2 |               block-1              |              rack-2              |
+| node-3 |               block-2              |              rack-1              |
+| node-4 |               block-2              |              rack-3              |
+
+Note that, there is a pair of nodes, node-1 and node-3, with the same value of
+the "cloud.provider.com/topology-rack" label, but in different blocks.
+
+### Admin-facing API
+
+```golang
+// ResourceFlavorSpec defines the desired state of the ResourceFlavor
+type ResourceFlavorSpec struct {
+    ...
+
+  // TopologyName indicates the name of the topology for the ResourceFlavor.
+  // When specified, it enables scraping of the topology information from the
+  // nodes matching to the Resource Flavor node labels.
+  //
+  // +optional
+  TopologyName *string `json:"topologyName,omitempty"`
+}
+
+// TopologySpec defines the desired state of Topology
+type TopologySpec struct {
+  // Levels defines the levels of topology.
+  //
+  // +listType=atomic
+  // +kubebuilder:validation:MinItems=1
+  // +kubebuilder:validation:MaxItems=5
+  Levels []TopologyLevel `json:"levels,omitempty"`
```

Comment:
Yes they get evaluated in order: from the one specified going up to the top. At each step we check if it is possible to satisfy the resource demand within a single topology domain. Say user specifies rack. Then we check rack, if not possible we fallback to block, if not possible we admit using multiple blocks.

Reply:
That makes sense. It would be great to mention the 3 step evaluation mechanism in API comment or dedicated section.

Additionally, I'm wondering if we need to evolve the internal PodSetInfo mechanism to support TAS.
Maybe we need to have nodeSelector information for every Pod instead of PodSet, right?

Reply:
> It would be great to mention the 3 step evaluation mechanism in API comment or dedicated section. 

Good point, I have added explanation in the documentation for the "preferred" annotation. PTAL.

Reply:
> I don't think we need to extend the PodSetInfo mechanism. This mechanism is used to restore the nodeSelectors (and other proporties of a Job) on suspend / eviction. However, we can restore it to the same values regardless of the assignment during admission.

That makes sense. Thank you for the clarifications.

> Also, the PodSetInfo is an internal structure, rather than part of the API so I think we can figure it out during the implementation phase, but I don't expect a need to change it.

Yeah, I think so too, I just cared what we will go with PodSetsinfo after we introduce this feature. I think we do not think that we need to mention that here as well.

Reply:
> > It would be great to mention the 3 step evaluation mechanism in API comment or dedicated section.
> 
> 
> Good point, I have added explanation in the documentation for the "preferred" annotation. PTAL.

Great to see. Thank you!